packages/xo-server/src/_MultiKeyMap.js

```diff
@@ -1,58 +1,88 @@
-const kValue = Symbol('value')
+class Node {
+  constructor(value) {
+    this.children = new Map()
+    this.value = value
+  }
+}
 
-function del(map, i, n, keys) {
+function del(node, i, n, keys) {
   if (i === n) {
-    return map.delete(kValue)
+    if (node instanceof Node) {
+      node.value = undefined
+      return node
+    }
+    return
+  }
+  if (!(node instanceof Node)) {
+    return node
   }
   const key = keys[i]
-  const child = map.get(key)
+  const { children } = node
+  const child = children.get(key)
   if (child === undefined) {
-    return false
+    return node
   }
-  const deleted = del(child, i + 1, n, keys)
-  if (child.size === 0) {
-    map.delete(key)
+  const newChild = del(child, i + 1, n, keys)
+  if (newChild === undefined) {
+    if (children.size === 1) {
+      return node.value
+    }
+    children.delete(key)
+  } else if (newChild !== child) {
+    children.set(key, newChild)
   }
-  return deleted
+  return node
 }
 
-function get(map, i, n, keys) {
-  if (i === n) {
-    return map.get(kValue)
-  }
-  const child = map.get(keys[i])
-  if (child !== undefined) {
-    return get(child, i + 1, n, keys)
-  }
+function get(node, i, n, keys) {
+  return node instanceof Node
+    ? i === n
+      ? node.value
+      : get(node.children.get(keys[i]), i + 1, n, keys)
+    : i === n
+    ? node
+    : undefined
 }
 
-function set(map, i, n, keys, value) {
+function set(node, i, n, keys, value) {
   if (i === n) {
-    return map.set(kValue, value)
+    if (node instanceof Node) {
+      node.value = value
+      return node
+    }
+    return value
   }
   const key = keys[i]
-  let child = map.get(key)
-  if (child === undefined) {
-    map.set(key, (child = new Map()))
+  if (!(node instanceof Node)) {
+    node = new Node(node)
+    node.children.set(key, set(undefined, i + 1, n, keys, value))
+  } else {
+    const { children } = node
+    const child = children.get(key)
+    const newChild = set(child, i + 1, n, keys, value)
+    if (newChild !== child) {
+      children.set(key, newChild)
+    }
   }
-  set(child, i + 1, n, keys, value)
+  return node
 }
 
 export default class MultiKeyMap {
   constructor() {
-    this._map = new Map()
+    // each node is either a value or a Node if it contains children
+    this._root = undefined
   }
 
   delete(keys) {
-    return del(this._map, 0, keys.length, keys)
+    this._root = del(this._root, 0, keys.length, keys)
```

Comment:
Why don't you return the new map here but you do in `set`? Or maybe this could return a boolean.